### ISSUE TITLE
Question regarding reproducibility of train, validation splits

### ISSUE THREAD

**Issue by alejandrodumas** [PyTorchLightning/pytorch-lightning-bolts#120]
Are the train, validations splits of the dataset guaranteed? From what I see, if I'm not 
missing something, randomness depends on [random_split](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/datamodules/mnist_datamodule.py#L99) that uses PyTorch `default_generator` if a new PyTorch version changes this, is it possible that 2 different versions of the MNISTDataModule for example with 2 torch versions have different splits?
Thanks and sorry if this is obvious!

**Comment by ananyahjha93** [comment]
@alejandrodumas hey, thanks for the question.

The main idea here is that the test set for these data modules are already separate. For datasets with an even distribution of examples across classes, having a random split for train-val should not matter. It also tests your model's generalization capabilities if you are training on separate train-val splits for different runs. You models should be able to give similar performance on random splits. (Also, for unsupervised learning, this should not be an issue at all)

However, we can look into this for unbalanced datasets which have limited examples for certain classes.

**Comment by alejandrodumas** [comment]
Thanks @ananyahjha93 for the clarification, I have one more doubt regarding this subject if you don't mind, if we read this [line](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/datamodules/mnist_datamodule.py#L77)  and compare it with [this](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/datamodules/mnist_datamodule.py#L99)  we call in two different instances the function `random_split` shouldn't the generation of the train and validation be related? Could it be the case that we're mixing examples that could be present on both classes? I think that we should store the indices of the train set so the validation set is aware. Thanks again!

**Comment by ananyahjha93** [comment]
@alejandrodumas looking into this!

**Comment by nateraw** [comment]
@alejandrodumas The splits you linked to will be encouraged to be done in the datamodule's `setup` hook, similar to how you would do something like that in a `LightningModule`. This way you'd only make one call instead of two as you showed above. Great catch on that btw 😄 .

The `LightningDataModule` will come packaged directly with lightning in version 0.9.0 and will be removed from bolts. You can find the docs on the latest version of the built-in datamodule class in the [lightning docs here.](https://pytorch-lightning.readthedocs.io/en/latest/datamodules.html)

@ananyahjha93 Let's close this once we remove calls to `random_split` from dataloader hooks across this repo (if you think that's reasonable). That way there is less confusion.

**Comment by williamFalcon** [comment]
@alejandrodumas , yes the random_split needs the generator arg to make it deterministic. I've updated those in bolts now. Otherwise you will have also trained on val data during training. Looks like pytorch 1.6 added that support.
